controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/InternalStepRunner.java

```diff
@@ -346,20 +347,48 @@ else if (timedOut(id, deployment.get(), endpointTimeout)) {
                 return Optional.of(error);
             }
         }
-        controller.jobController().locked(id, lockedRun -> lockedRun.withSummary(summary));
 
-        if (timedOut(id, deployment.get(), installationTimeout)) {
+        boolean failed = false;
+
+        NodeList suspendedTooLong = nodeList.suspendedSince(controller.clock().instant().minus(installationTimeout));
+        if ( ! suspendedTooLong.isEmpty()) {
+            logger.log(INFO, "Some nodes have been suspended for more than " + installationTimeout.toMinutes() + " minutes.");
+            failed = true;
+        }
+
+        if (run.noNodesDownSince()
+               .map(since -> since.isBefore(controller.clock().instant().minus(installationTimeout)))
+               .orElse(false)) {
+            if (summary.needPlatformUpgrade() > 0 || summary.needReboot() > 0 || summary.needRestart() > 0)
+                logger.log(INFO, "No nodes allowed to suspend to progress installation for " + installationTimeout.toMinutes() + " minutes.");
+            else
+                logger.log(INFO, "Nodes not able to start with new application package.");
+            failed = true;
+        }
+
+        Duration timeout = JobRunner.jobTimeout.minusHours(1); // Time out before job dies.
+        if (timedOut(id, deployment.get(), timeout)) {
+            logger.log(INFO, "Installation failed to complete within " + timeout.toHours() + "hours!");
```

Comment:
Should probably add some advise on what the customer should do next when the deployment fails?

Reply:
Yes.